static/js/message_fetch.js

```diff
@@ -49,16 +49,13 @@ function process_result(data, opts) {
 
     messages = messages.map((message) => message_helper.process_new_message(message));
 
-    // In case any of the newly fetched messages are new, add them to
-    // our unread data structures.  It's important that this run even
-    // when fetching in a narrow, since we might return unread
-    // messages that aren't in the home view data set (e.g. on a muted
-    // stream).
-    //
-    // BUG: This code path calls pm_list.update_private_messages, even
-    // if there were no private messages (or even no new messages at
-    // all) in data.messages, which is a waste of resources.
-    message_util.do_unread_count_updates(messages);
+    // We track all unread messages sent to us in `page_params.unread_msgs`
+    // by server and update UI for any change in unread messages via unread.js.
+    // We can still get new unread messages that we don't track if user has 50k+ unread
+    // messages (See MAX_UNREAD_MESSAGE in zerver/lib/message.py).
+    if (page_params.unread_msgs.old_unreads_missing) {
+        message_util.do_unread_count_updates(messages);
+    }
```

Comment:
This looks correct; opened https://github.com/zulip/zulip/issues/23334 for a follow-up cleanup inspired by reading this.

Reply:
I did a bit more thinking and I think an architecture that pushes this check lower in the stack would be more robust at catching possible bugs; I'll push a revision.